File: java/TrappingRainWater.java
public int trap(int[] height) {
        if(height.length < 3) return 0;
        int i = 0, j = height.length - 1, res = 0, plank = 0;
        while(i <= j) {
            plank = Math.max(plank, Math.min(height[i], height[j]));
            res = height[i] >= height[j]? res + (plank - height[j--]) : res + (plank - height[i++]);
        }
        return res;
    }

File: Array/trappingRainWater.kt
class Solution {
    fun trap(height: IntArray): Int {
        // O(n) time, O(1) space
        val n = height.size
        var res = 0
        var l = 0
        var maxL = height[l]
        var r = n-1
        var maxR = height[r]

        while (l < r) {
            if (maxL <= maxR) {
                l++
                maxL = maxOf(maxL, height[l])
                res+= maxL - height[l]
            } else {
                r--
                maxR = maxOf(maxR, height[r])
                res+= maxR - height[r]
            }
        }

        return res
    }
}
